server/application-server/src/main/java/de/tum/in/www1/hephaestus/leaderboard/SlackMessageService.java

```diff
@@ -0,0 +1,165 @@
+
+    @Value("${slack.channelId}")
+    private String channelId;
+
+    @Value("${slack.runScheduledMessage}")
+    private boolean runScheduledMessage;
+
+    @Value("${spring.url:localhost:8080}")
+    private String hephaestusUrl;
+
+    @Autowired
+    private App slackApp;
+
+    @Autowired
+    private LeaderboardService leaderboardService;
+
+    public void sendMessage(String channelId, List<LayoutBlock> blocks, String fallback)
+            throws IOException, SlackApiException {
+        ChatPostMessageRequest request = ChatPostMessageRequest.builder()
+                .channel(channelId)
+                .blocks(blocks)
+                .text(fallback) // used for example in notifications
+                .build();
+
+        ChatPostMessageResponse response = slackApp.client().chatPostMessage(request);
+
+        if (!response.isOk()) {
+            logger.error("Failed to send message to Slack channel: " + response.getError());
+        }
+    }
+
+    private List<User> getTop3SlackReviewers() {
+        // exactly 7 days ago
+        OffsetDateTime after = OffsetDateTime.of(LocalDate.now().minusDays(7), OffsetDateTime.now().toLocalTime(),
+                OffsetDateTime.now().getOffset());
+        var leaderboard = leaderboardService.createLeaderboard(Optional.of(after), Optional.empty(),
+                Optional.empty());
+        var top3 = leaderboard.subList(0, Math.min(3, leaderboard.size()));
+        logger.debug("Top 3 Users of the last week: " + top3.stream().map(e -> e.user().name()).toList());
+
+        List<User> allSlackUsers;
+        try {
+            allSlackUsers = slackApp.client().usersList(r -> r).getMembers();
+        } catch (SlackApiException | IOException e) {
+            logger.error("Failed to get Slack users: " + e.getMessage());
+            return new ArrayList<>();
+        }
+
+        return top3.stream().map(entry -> {
+            var exactUser = allSlackUsers.stream()
+                .filter(user -> user.getName().equals(entry.user().name()) || (user.getProfile().getEmail() != null && user.getProfile().getEmail().equals(entry.user().email()))).findFirst();
+            if (exactUser.isPresent()) {
+                return exactUser.get();
+            }
+
+            // find through String edit distance
+            return allSlackUsers.stream()
+                .min((a, b) -> Integer.compare(
+                    LevenshteinDistance.getDefaultInstance().apply(
+                        entry.user().name(), a.getName()),
+                    LevenshteinDistance.getDefaultInstance().apply(
+                        entry.user().name(), b.getName()))).orElse(null);
+        }).filter(user -> user != null).toList();
+    }
+
+    @Scheduled(cron = "${slack.cronSchedule}")
+    public void sendScheduledLeaderboard() {
+        if (!runScheduledMessage) {
+            return;
+        }
+
+        // get date in unix format
+        var currentDate = OffsetDateTime.now().toEpochSecond();
```

Comment:
We should use exact times for the top3 and the leaderboard calculation

Reply:
What do you mean by "exact times"?

`currentDate` is only used for the date-part of the message:
![image](https://github.com/user-attachments/assets/52bee2e6-565c-4a0e-8e1c-de223269f5d1)

Reply:
I guess I'm just worried that this is running while we do a github sync. In that case the main thread is occupied, but this is something we anyways have to fix.

Reply:
Also we have the exact times and don't need to use `now` everywhere

Reply:
I personally didn't even consider this, but I totally understand your concern.
In my new approach with the `ThreadPoolTaskScheduler` this is now solved as well since we use the exact time.